File: C/c_solutions_21-30/Euler_24.c
/* 
 * What is the millionth lexicographic permutation of the digits
 * 0, 1, 2, 3, 4, 5, 6, 7, 8, and 9?
 *
 */

#include "euler_util.h"

#define PERMUTATION_SIZE 10 
#define PERMUTATION_LIMIT 1000000


void sort(int permutation[], int idx)
{
    int i, j, curr;
    for (i=idx; i < PERMUTATION_SIZE - 1; i++) {
        curr = permutation[i + 1];
        j = i;
        while (j > idx && curr < permutation[j]) {
            permutation[j + 1] = permutation[j];
            permutation[j] = curr;
            j--;
        }
        permutation[j + 1] = curr;
    }
}

void swap(int permutation[], int idx1, int idx2)
{
    int tmp;
    tmp = permutation[idx1];
    permutation[idx1] = permutation[idx2];
    permutation[idx2] = tmp;    
}

int next_largest_idx(int permutation[])
{
    int i;
    for (i=PERMUTATION_SIZE - 1; i > 0; i--)
        if (permutation[i] > permutation[i - 1])
            return i - 1;
    return i;
}

int last_largest_idx(int permutation[], int idx)
{
    int i;
    for (i=PERMUTATION_SIZE - 1; i >= idx; i--)
        if (permutation[i] > permutation[idx])
            return i;
    return i;
}

int main(int argc, char *argv[])
{
    float start = timeit();

    int one_to_ten[PERMUTATION_SIZE] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    int first_idx, second_idx, count;

    int i;

    for (count=0; count < PERMUTATION_LIMIT - 1; count++) {
        first_idx = next_largest_idx(one_to_ten);
        second_idx = last_largest_idx(one_to_ten, first_idx);
        swap(one_to_ten, first_idx, second_idx);
        sort(one_to_ten, first_idx);
    }

    float stop = timeit();
    
    printf("Answer: ");

    for (i=0; i < PERMUTATION_SIZE; i++)
        printf("%d", one_to_ten[i]);

    printf("\nTime: %f\n", stop - start);

    return 0;
}

File: Java/java_solutions_21-30/Euler_24.java
/*
/   Project Euler problem #24 Lexicographic permutations
/
/   What is the millionth lexicographic permutation of the digits 
/   0, 1, 2, 3, 4, 5, 6, 7, 8, and 9?
/
*/

public class Euler_24
{
    public static String permutationSort(char[] permutationArray, int idx)
    {
        for (int i = idx; i < permutationArray.length; i++) {

            char cur = permutationArray[i];
            int j = i - 1;

            while (j >= idx && cur < permutationArray[j]) 
                permutationArray[j + 1] = permutationArray[j--];

            permutationArray[j + 1] = cur;
        }

        return new String(permutationArray);
    }

    public static String permute(String permutationString)
    {
        char[] permutationArray = permutationString.toCharArray();

        int idx1, idx2;

        for (idx1 = permutationArray.length - 1; idx1 > 0; idx1--) {
            if (permutationArray[idx1] > permutationArray[idx1 - 1]) {
                idx1--;
                break;
            }
        }

        for (idx2 = permutationArray.length - 1; idx2 > idx1; idx2--) {
            if (permutationArray[idx1] < permutationArray[idx2]) {
                char tmp = permutationArray[idx1];
                permutationArray[idx1] = permutationArray[idx2];
                permutationArray[idx2] = tmp;
                return permutationSort(permutationArray, idx1 + 1);
            }
        }

        char tmp = permutationArray[idx1];
        permutationArray[idx1] = permutationArray[idx2];
        permutationArray[idx2] = tmp;
        return permutationSort(permutationArray, idx1 + 1);
    }

    public static void main(String[] args)
    {
        String permutationString = "0123456789";
        for (int i = 1; i < 1000000; i++)
            permutationString = permute(permutationString);

        System.out.println("Answer: " + permutationString);
    }
}
